vsifile/io/base.py

```diff
@@ -70,34 +71,36 @@ def __hash__(self):
         ...
 
     def _get_header(self) -> bytes:
-        logger.debug("VSIFILE_INFO: HEAD (Open)")
-        head = obs.head(self._store, self._key)
-        self._size = head["size"]
-
-        header = self.header_cache.get(f"{self.name}-header", read=True)
-        if not header:
-            end = (
-                vsi_settings.ingested_bytes_at_open
-                if self._size > vsi_settings.ingested_bytes_at_open
-                else self._size
-            )
+        cache = self.header_cache.get(f"{self.name}-header", read=True)
+        if not cache:
             logger.debug("VSIFILE_INFO: GET")
-            logger.debug(f"VSIFILE: Downloading: 0-{end}")
-            header = bytes(obs.get_range(self._store, self._key, start=0, end=end))
+            logger.debug(f"VSIFILE: Downloading: 0-{vsi_settings.ingested_bytes_at_open}")
+            response = obs.get(
+                self._store,
+                self._key,
+                options={"range": (0, vsi_settings.ingested_bytes_at_open)},
+            )
+
+            meta = response.meta
+            header = bytes(response.bytes())
```

Comment:
This line seems to take ~0.1 seconds 😬 

`response.bytes()` returns something of type `builtins.Bytes` which needs to be converted to `bytes` in order to be stored in the cache 🤷‍♂️ 

cc @kylebarron

Reply:
You should split that into two lines.

```
response.bytes()
```
downloads the content of the request into a byte buffer, and could be where the majority of your time is coming from. (This should be considered part of the request time)

Then passing that to `bytes()` should be a fast memory copy.

> needs to be converted to `bytes` in order to be stored in the cache 🤷‍♂️

Why do you need to do that? Or do you mean it needs to be hashable? We could make `Bytes` hashable

> type `builtins.Bytes`

We should fix that `__module__` so that it appears to be from `obstore.Bytes`

Reply:
> Why do you need to do that? Or do you mean it needs to be hashable? We could make Bytes hashable

👇 is what I'm getting when trying to set into the cache 

> TypeError: cannot pickle 'builtins.Bytes' object

But anyway I'm pretty sure I need to pass the `bytes` to rasterio

Reply:
> > Why do you need to do that? Or do you mean it needs to be hashable? We could make Bytes hashable
> 
> 👇 is what I'm getting when trying to set into the cache
> 
> > TypeError: cannot pickle 'builtins.Bytes' object

Can you point out what line that is? What kind of cache are you using? It seems weird that you'd want to pickle a buffer... A buffer is already serialized!

> But anyway I'm pretty sure I need to pass the `bytes` to rasterio

You most likely don't need to convert the `obstore.Bytes` class to a stdlib `bytes` object. rasterio most likely supports the buffer protocol natively, in which case it can read directly from rust's memory.

Reply:
https://github.com/grantjenks/python-diskcache/blob/master/diskcache/core.py#L221

maybe I can play with the `pickle_protocol` option

Reply:
quick replicable test

```python
from obstore.store import S3Store
import obstore as obs
import pickle

store = S3Store("sentinel-cogs", config={"AWS_REGION": "us-west-2", "skip_signature": True})
response = obs.get(store, "sentinel-s2-l2a-cogs/15/T/VK/2023/10/S2B_15TVK_20231008_0_L2A/TCI.tif", options={"range": (0, 10)})

pickle.dumps(response.bytes())
```

Reply:
@kylebarron when I try passing  obstore.Bytes to rasterio I'm getting a system crash ;-) 

So I'm guess I'm fine doing response.bytes().to_bytes()

Reply:
Hmm, we can enable pickling for `Bytes`. That's not hard to do. 

> @kylebarron when I try passing obstore.Bytes to rasterio I'm getting a system crash ;-)

Hmm. I'm curious which side that's from. Given that we can pass `obstore.Bytes` to `numpy` directly (and I haven't seen a crash from that yet), I'd like to think it's rasterio's problem 🤷‍♂️